keps/sig-storage/661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,1112 @@
+
+  
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Before we proceed to the proposal it is imperative to layout the various components involved
+in creating and managing the persistent volumes associated with `StatefulSet`. 
+
+- StatefulSet controller creates the PVCs from `volumeClaimTemplates` and associates them with the pods
+spawned from it. Reconciliation logic runs for every change made to the StatefulSet
+`Spec` and `Status` reflects the progress towards making the real world state equal to the
+`Spec`.
+- The validation logic in the API server is responsible for restricting the fields
+which can be edited in a StatefulSet specification.
+  - Part of the validation that the API server performs for objects (eg `StatefulSet`, `PVC`) happens in the object's API server code and part happens in an admission plugin.
+- The RBAC specification in the controller policy is responsible for adding the required 
+RBAC permission for StatefulSet.
+- The StatefulSet controller also supports revision control.
+
+The following changes are proposed to achieve the goal:
+1. Make the `volumeClaimTemplates` storage size (`.spec.resources.request.storage`) editable by modification to the api server
+   validation. We should allow the user to both increase or decrease the storage size in case of expansion failures. Allowing the user to decrease the storage size will give them the ability to address failed expansions, as described in [KEP 1790](../1790-recover-resize-failure/README.md).
+2. Add reconciliation of the associated PVC's size and the individual `volumeClaimTemplates` size into the 
+   StatefulSet controller reconciliation.
+3. Add PVC `patch` privilege to the StatefulSet controller's `ClusterRole` RBAC.
+4. Modification to the `StatefulSet.Status` to add additional field for `volumeClaimTemplates` to track volume expansion workflow status
+
+Once the above changes are made, modification to any of the `volumeClaimTemplates` storage
+size (specifically to the `.spec.resources.request.storage` field) by user will update (patch) the underlying PVC via the reconciliation performed by (2) above.
+The existing PVC resize workflow kicks in at this point and attempts to resize the volume.
+
+![Diagram 1](./diag1.png)
+
+Two things to note about the diagram above
+* It's an over simplification of the volume resize process or of how StatefulSet reconciliation works - but it should suffice in the context of this discussion.
+* The numbers I denote on the arrows do not indicate the order in which operations happen. They are there for easier identification.
+
+### Possible Errors/Failures
```

Comment:
It would be good to match the current KEP template, see https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/README.md Here you should start the `Risks and Mitigations` section.